examples/go_router/lib/router.dart

```diff
@@ -0,0 +1,74 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:go_router/go_router.dart';
+
+import 'main.dart';
+import 'user.dart';
+
+final routerProvider = Provider<GoRouter>((ref) {
+  return GoRouter(
+    // Keep this to `true` if want to know what's going on under the hood
+    debugLogDiagnostics: true,
+    redirect: (state) {
+      // We want to READ the state, here.
+      // GoRouter is already aware of state changes through `refreshListenable`
+      final user = ref.read(userProvider);
+
+      // From here we can use the state and implement our custom logic
+      final areWeLoggingIn = state.location == '/login';
+
+      if (user == null) {
+        // We're not logged in
+        // So, IF we aren't in the login page, go there.
+        return areWeLoggingIn ? null : '/login';
+      }
+      // We're logged in
+
+      // At this point, IF we're in the login page, go to the home page
+      if (areWeLoggingIn) return '/';
+
+      // There's no need for a redirect at this point.
+      return null;
+    },
+    // This is crucial to make the router work with Riverpod.
+    refreshListenable: RouterNotifier(ref),
```

Comment:
@lucavenir 

This can be written like this by using [GoRouterRefreshStream](https://pub.dev/documentation/go_router/latest/go_router/GoRouterRefreshStream-class.html) without RouterNotifier:

```dart
  final refreshListenable = GoRouterRefreshStream(
    ref.watch(userProvider.notifier).stream,
  );
  ref.onDispose(refreshListenable.dispose);
  return GoRouter(
    refreshListenable: refreshListenable,
```

Reply:
Hi!

Thanks!! This would surely simplify things. The reason I built a `RouterNotifier` is because I wanted to "emulate" or "be as close as possible" to the `StateNotifier` class which is core in this repo.

Furthermore, doesn't `GoRouterRefreshStream` listen just one stream at a time? While it is surely possible to build a `stream` out of many providers, I think that adding a simple `listen` for each provider needed to contribute to `GoRouter` 's state is simpler.

Let me know what you think about this